server/datastore/mysql/hosts.go

```diff
@@ -1223,13 +1222,16 @@ func (ds *Datastore) applyHostFilters(
 				//                      but either way we haven't canceled it.
 				// bsehr.error IS NULL <- this means the batch script framework didn't mark this host as incompatible
 				//                        with this script run.
-				batchScriptExecutionIDFilter += ` AND (hsr.exit_code IS NULL AND (hsr.canceled IS NULL OR hsr.canceled = 0) AND bsehr.error IS NULL)`
+				batchScriptExecutionIDFilter += ` AND ((hsr.host_id AND (hsr.exit_code IS NULL AND (hsr.canceled IS NULL OR hsr.canceled = 0) AND bsehr.error IS NULL)) OR (hsr.host_id is NULL AND ba.canceled = 0 AND bsehr.error IS NULL))`
```

Comment:
A host can be pending if:

The batch is running, so the host has a host_script_results record (hsr.host_id is not null) with no exit code (so it hasn't run yet)
The batch hasn't started yet, so the host has no host_script_results record (hsr.host_is is null), and the batch isn't canceled (ba.canceled = 0)